File: generators/kotlin_ktor_client_generator/test/src/jvmTest/kotlin/com/example/kitchensink/client/GenerationErrorsTest.kt
package com.example.kitchensink.client

import java.io.File
import java.nio.file.Files
import kotlin.test.Test
import kotlin.test.assertFalse
import kotlin.test.assertTrue

// Shells out to the openapi-yagen binary directly (not through Gradle's own generateClient task,
// which regenerates the *happy-path* kitchensink.yaml fixture under default strict=true) to prove
// an unsupported request-body content-type actually fails generation loudly instead of silently
// dropping the body - see resources/unsupported_content_type.yaml and lib/operations.js's
// pickBodyContent/buildRequestBody. Paths come from system properties set by build.gradle.kts's
// own tasks.withType<Test> block, not guessed from the JVM test process's working directory.
class GenerationErrorsTest {
    private val bin = System.getProperty("openApiYagenBin")
    private val generatorSrc = System.getProperty("generatorSrcDir")
    private val spec = File(System.getProperty("testResourcesDir"), "unsupported_content_type.yaml").absolutePath

    private fun generate(outDir: File, vararg extraArgs: String): Pair<Int, String> {
        val command = mutableListOf(
            bin, "g", "-o", outDir.absolutePath, "-g", generatorSrc, "-c", spec,
            "-v", "packageName=com.example.unsupported"
        )
        command.addAll(extraArgs)
        val process = ProcessBuilder(command).redirectErrorStream(true).start()
        val output = process.inputStream.bufferedReader().readText()
        val exitCode = process.waitFor()
        return exitCode to output
    }

    @Test
    fun `an unsupported request body content-type aborts generation by default (strict=true)`() {
        val outDir = Files.createTempDirectory("kotlin-unsupported-strict-").toFile()
        try {
            val (exitCode, output) = generate(outDir)
            assertTrue(exitCode != 0, "expected generation to fail in strict mode (default): $output")
            assertTrue(output.contains("text/plain"), output)
        } finally {
            outDir.deleteRecursively()
        }
    }

    @Test
    fun `an unsupported request body content-type is skipped with a warning under strict=false`() {
        val outDir = Files.createTempDirectory("kotlin-unsupported-permissive-").toFile()
        try {
            val (exitCode, output) = generate(outDir, "-v", "strict=false")
            assertTrue(exitCode == 0, "expected generation to succeed under -v strict=false: $output")
            assertTrue(output.contains("WARNING"), output)
            assertTrue(output.contains("text/plain"), output)
            assertFalse(File(outDir, "apis/NotesApi.kt").exists())
        } finally {
            outDir.deleteRecursively()
        }
    }
}

File: generators/go_net_http_client_generator/test/generation_errors_test.go
package clienttest

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
)

// Shells out to the openapi-yagen binary directly (not through generate.sh, which regenerates the
// *happy-path* kitchensink.yaml fixture under default strict=true) to prove an unsupported query
// parameter array serialization style actually fails generation loudly instead of silently
// emitting the wrong wire format - see resources/unsupported_query_array_style.yaml and
// operations.js's buildArrayQueryParam.

func openApiYagenBinary() string {
	if bin := os.Getenv("OPENAPI_YAGEN"); bin != "" {
		return bin
	}
	return "../../../dist/openapi-yagen"
}

func generateQueryArrayStyleFixture(t *testing.T, outDir string, extraArgs ...string) (string, error) {
	t.Helper()
	args := []string{
		"g", "-o", outDir,
		"-g", "../src",
		"resources/unsupported_query_array_style.yaml",
		"-v", "packageName=go_net_http_client_generator_test/unsupported",
	}
	args = append(args, extraArgs...)
	cmd := exec.Command(openApiYagenBinary(), args...)
	out, err := cmd.CombinedOutput()
	return string(out), err
}

func TestUnsupportedQueryArrayStyleAbortsGenerationByDefault(t *testing.T) {
	outDir := t.TempDir()
	output, err := generateQueryArrayStyleFixture(t, outDir)
	if err == nil {
		t.Fatalf("expected generation to fail in strict mode (default), output:\n%s", output)
	}
	if !strings.Contains(output, "matrix") {
		t.Fatalf("expected error output to mention the unsupported style, got:\n%s", output)
	}
}

func TestUnsupportedQueryArrayStyleIsSkippedWithAWarningUnderStrictFalse(t *testing.T) {
	outDir := t.TempDir()
	output, err := generateQueryArrayStyleFixture(t, outDir, "-v", "strict=false")
	if err != nil {
		t.Fatalf("expected generation to succeed under -v strict=false, output:\n%s", output)
	}
	if !strings.Contains(output, "WARNING") || !strings.Contains(output, "matrix") {
		t.Fatalf("expected a WARNING mentioning the unsupported style, got:\n%s", output)
	}
	if _, err := os.Stat(filepath.Join(outDir, "client", "NotesClient.go")); !os.IsNotExist(err) {
		t.Fatalf("the only operation on the only tag was skipped, so no client file should exist for it (stat err: %v)", err)
	}
}
